sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -120,7 +120,25 @@ case class Generate(
 
 case class Filter(condition: Expression, child: LogicalPlan)
   extends UnaryNode with PredicateHelper {
-  override def output: Seq[Attribute] = child.output
+
+  // The columns that will filtered out by `IsNotNull` could be considered as not nullable.
+  private lazy val notNullAttributes = splitConjunctivePredicates(condition).flatMap {
```

Comment:
Why using `condition`, instead of `constraints`

Reply:
IIUC `condition` has all the `constrains` in this node? `InferFiltersFromConstraints` updates `condition` based on `constrains`.